pecos/core/utils/mmap_util.hpp

```diff
@@ -0,0 +1,509 @@
+                }
+                // Load meta offset
+                meta_offset = file_util::fget_one<uint64_t>(fp);
+            }
+
+            /* Obtain current machine endianness */
+            uint8_t get_endianess() const {
+                return static_cast<uint8_t>(file_util::runtime());
+            }
+        };
+
+        std::vector<MetaInfo> _info; // offset and size of each block in the file
+        uint64_t _iter = 0; // Iterator index for retrieving metadata info in blocks order
+}; // end class _MmapFileMetadata
+
+
+/*
+ * Module Private Class to load data from a memory-mapped file.
+ */
+class _MmapFileLoad {
+    public:
+        /* Constructor loads from memory-mapped file name
+         * Parameters:
+         *      file_name (const std::string &): Name of file to load from.
+         *      pre_load (const bool): Whether to pre-fault all pages into memory before accessing.
+         */
+        _MmapFileLoad(const std::string & file_name, const bool pre_load) {
+            // Load metadata first
+            FILE * fp = fopen(file_name.c_str(), "rb");
+            if (!fp) {
+                throw std::runtime_error("Load metadata: Open file failed.");
+            }
+            _metadata.load(fp);
+            if (fclose(fp) != 0) {
+                throw std::runtime_error("Load metadata: Close file failed.");
+            }
+
+            // Load mmap
+            int fd = open(file_name.c_str(), O_RDONLY);
+            if (fd == -1) {
+                throw std::runtime_error("Load Mmap file: Open file failed.");
+            }
+            _load_mmap(fd, pre_load);
+            if (close(fd) < 0) {
+                throw std::runtime_error("Load Mmap file: Close file failed.");
+            }
+        }
+        /* Destructor frees the memory-mapped region */
+        ~_MmapFileLoad() { _free_mmap(); }
+
+        /* Load a simple-serializable vector(array)
+         * Parameters:
+         *      n_elements (uint64_t): Number of elements to load for the array.
+         */
+        template<class T, class TT=T, if_simple_serializable<TT> = true>
+        T * fget_multiple(uint64_t n_elements) {
+            auto meta_info = _metadata.next();
+
+            // Verification
+            if (n_elements * static_cast<uint64_t>(sizeof(T)) != meta_info.size) {
+                throw std::runtime_error(
+                    "This block contains " + std::to_string(meta_info.size) +
+                    " bytes data, retrieving " + std::to_string(n_elements * static_cast<uint64_t>(sizeof(T))) +
+                    " bytes not equal. Please double check.");
+            }
+
+            return reinterpret_cast< T * >(reinterpret_cast< char * >(_mmap_ptr) + meta_info.offset);
+        }
+
+        /* Load one element */
+        template<class T>
+        T fget_one() { // Call fget_multiple
+            return * fget_multiple<T>(1u);
```

Comment:
There is no implementation of fget_multiple<T>() when T is not  simple_serializable.